### PR TITLE
Block API: Fix the block 'edit' property validation

### PR DESCRIPTION
**Pull request by Mamaduka**
## What?
PR fixes the block's `edit` property validation when bootstrapping the block types, failing for memoized components.

## Why?
The main requirement for the `edit` property is to be a valid React component for the block editor to render, which usually means having the type of a `function`, but can be a special `object` when wrapped in React APIs.

I noticed this while working on #57173. The HoCs like `withSelect` will return `pure` or memoized components.

## How?
Use the [`isValidElementType`](https://github.com/facebook/react/blob/main/packages/shared/isValidElementType.js#L38) from the official [`react-is`](https://www.npmjs.com/package/react-is) library to check the `edit` property's validity correctly.

P.S. If this check is too extensive, we can create our custom utility using primitives provided by the library.

## Testing Instructions
1. Apply the patch provided below.
2. Open the post editor.
3. It will `The "edit" property must be a valid function.` error on the trunk.
4. Repeat the same step on this branch.
5. There should be no errors in the console.

### Patch
```diff
diff --git packages/block-library/src/paragraph/edit.js packages/block-library/src/paragraph/edit.js
index ac766f69dd8..43e40b3bc0e 100644
--- packages/block-library/src/paragraph/edit.js
+++ packages/block-library/src/paragraph/edit.js
@@ -22,6 +22,7 @@ import {
 } from '@wordpress/block-editor';
 import { createBlock } from '@wordpress/blocks';
 import { formatLtr } from '@wordpress/icons';
+import { memo } from '@wordpress/element';
 
 /**
  * Internal dependencies
@@ -183,4 +184,4 @@ function ParagraphBlock( {
 	);
 }
 
-export default ParagraphBlock;
+export default memo( ParagraphBlock );
```

### Testing Instructions for Keyboard
Same

### CODE REVIEW

File: packages/blocks/src/api/test/registration.js [hunk 1]
```diff
@@ -168,7 +168,7 @@ describe( 'blocks', () => {
 		it( 'should reject blocks with an invalid edit function', () => {
 			const blockType = {
 					save: noop,
-					edit: 'not-a-function',
```
Comment (Mamaduka) [inline packages/blocks/src/api/test/registration.js:171]: We can explicitly disallow strings, which is a valid element type for React. WDYT?

**Review by youknowriad — APPROVED** [review]
WFM

File: packages/blocks/package.json [hunk 2]
```diff
@@ -50,6 +50,7 @@
 		"hpq": "^1.3.0",
 		"is-plain-object": "^5.0.0",
 		"memize": "^2.1.0",
+		"react-is": "18.2.0",
```
Comment (gziolo) [inline packages/blocks/package.json]: We shouldn't pin the version here.

**Review by gziolo — APPROVED** [review]
Yes, looks good and tree-shaking ensures that only the used utils is bundled.

Let's make sure we don't pin the version of the dependency in the blocks package.

File: packages/blocks/package.json (hunk 2 above)
Comment (Mamaduka) [reply packages/blocks/package.json]: React version are pinned, so I used same logic here.

Comment (gziolo) [reply packages/blocks/package.json]: This is what I see:
Screenshot 2023-12-20 at 13 14 32

What do I miss here in regards to pinned React version?

**Review by tyxla — APPROVED** [review]
Nice find 👍

File: packages/blocks/package.json (hunk 2 above)
Comment (Mamaduka) [reply packages/blocks/package.json]: CleanShot 2023-12-21 at 10 23 46

It is pinned in the main package file.

Comment (gziolo) [reply packages/blocks/package.json]: Yes, we pin versions for the project. However, we should do our best to never pin them in WordPress packages.

**Comment by gziolo** [comment]
Nice, all goo to land it 🚀

**Pull request merged**